Create a quantized non-in-palce version CUDA ReLU function,

fufeisi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85669
Summary:
this and #85670 are to allow the relu function to run on a quantized tensor on cuda. That is torch.relu(qa) for a quantized tensor qa on cuda.
Test Plan:
python test/test_quantization.py

Previous PR that has been reverted: #85502.

fufeisi committed:
Create a quantized non-in-palce version CUDA ReLU function,
relu_quantized_cuda.

[ghstack-poisoned]

fufeisi committed:
Update on "Create a quantized non-in-palce version CUDA ReLU function,"


relu_quantized_cuda.

[ghstack-poisoned]

dzdang commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -253,7 +253,29 @@ def test_qrelu(self, X):
             }
         ]
-        self._test_activation_function(X, 'relu', relu_test_configs)
+        devices = ["cpu", "cuda"] if TEST_CUDA else ["cpu"]
```
let's use `get_supported_device_types` https://github.com/pytorch/pytorch/pull/79607/files

dzdang reviewed (approved):
I think we already approved these change in the previous PR. let's link that here to the summary body to preserve review history

z-a-f commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -253,7 +253,29 @@ def test_qrelu(self, X):
+            # Only test the non-in-place version relu quantized cuda,
+            # will remove this when creating in-place version relu quantized cuda.
+            if device == 'cuda':
+                relu_test_configs = [
+                    {
+                        'quantized_fn': [
+                            torch.relu,
+                            torch.nn.functional.relu,
+                        ],
+                        'reference_fn': torch.nn.functional.relu
+                    },
+                ]
```
nit: FYI, this is not a right way to "temporarily" override the configurations. The problem is that `relu_test_configs` is defined for the whole test. However, within the loop you are overriding it, meaning on all next iterations it will be the "overriden" version.

It's OK for this PR, as the next one removes these lines.

fufeisi replied:
OK, will avoid this next time.

z-a-f commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -253,7 +253,29 @@ def test_qrelu(self, X):
+                        'reference_fn': torch.nn.functional.relu
+                    },
+                ]
+            shapes = ((4,), (4, 4), (4, 4, 4), (4, 4, 4, 4))
+            dtypes = (torch.quint8, torch.qint8)
+            scales = (0.05, 0.1)
+            zero_points = (0, 5)
+            test_cases = itertools.product(shapes, dtypes, scales, zero_points)
+            for shape, dtype, scale, zero_point in test_cases:
+                X = torch.randn(*shape, device=device)
+                X = (X, (scale, zero_point, dtype))
+                self._test_activation_function(X, 'relu', relu_test_configs)
```
Is there a reason why you are removing he hypothesis, and manually writing the tests? Although it is one of the plans, it should not be done within the PR that introduces a new feature.

fufeisi replied:
If we keep the he hypothesis, assuming we are doing tests on both 'cpu' and 'cuda', the first iteration, we need to cast a NumPy array to a 'cpu' Tensor, and the second time, we need to either send the tensor to 'cuda' or keep the original NumPy array and cast it to 'cuda'. I think both ways will make the code longer.

z-a-f reviewed (changes requested):
We should not remove hypothesis in the same PR that introduces new features.

fufeisi committed:
Update on "Create a quantized non-in-palce version CUDA ReLU function,"

Summary:
this and #85670 are to allow the relu function to run on a quantized tensor on cuda. That is torch.relu(qa) for a quantized tensor qa on cuda.
Test Plan:
python test/test_quantization.py

Previous PR that has been reverted: #85502. 

[ghstack-poisoned]

fufeisi committed:
Update on "Create a quantized non-in-palce version CUDA ReLU function,"

Summary:
this and #85670 are to allow the relu function to run on a quantized tensor on cuda. That is torch.relu(qa) for a quantized tensor qa on cuda.
Test Plan:
python test/test_quantization.py

Previous PR that has been reverted: #85502. 

[ghstack-poisoned]

fufeisi commented:
@pytorchbot rebase

fufeisi committed:
Update on "Create a quantized non-in-palce version CUDA ReLU function,"

Summary:
this and #85670 are to allow the relu function to run on a quantized tensor on cuda. That is torch.relu(qa) for a quantized tensor qa on cuda.
Test Plan:
python test/test_quantization.py

Previous PR that has been reverted: #85502. 

[ghstack-poisoned]

fufeisi commented:
@pytorchbot merge
